moquette-0.17/broker/src/main/java/io/moquette/broker/unsafequeues/QueuePool.java

```diff
@@ -0,0 +1,490 @@
+
+        LinkedList<SegmentRef> acc = new LinkedList<>();
+        for (String segment : segments) {
+            final String[] split = segment.split(",");
+            final int idPage = Integer.parseInt(split[0].trim());
+            final int offset = Integer.parseInt(split[1].trim());
+
+            acc.offer(new SegmentRef(idPage, offset));
+        }
+        return acc;
+    }
+
+    public Queue getOrCreate(String queueName) throws QueueException {
+        final QueueName queueN = new QueueName(queueName);
+        if (queues.containsKey(queueN)) {
+            return queues.get(queueN);
+        } else {
+            // create new queue with first empty segment
+            final Segment segment = nextFreeSegment();
+            //notify segment creation for queue in queue pool
+            segmentedCreated(queueName, segment);
+
+            // When a segment is freshly created the head must the last occupied byte,
+            // so can't be the start of a segment, but one position before, or in case
+            // of a new page, -1
+            final Queue queue = new Queue(queueName, segment, VirtualPointer.buildUntouched(), segment, VirtualPointer.buildUntouched(),
+                this.allocator, callback, this);
+            queues.put(queueN, queue);
+            return queue;
+        }
+    }
+
+    /**
+     * Free mapped files
+     * */
+    public void close() throws QueueException {
+        allocator.close();
+
+        //save all into the checkpoint file
+        Properties checkpoint = new Properties();
+        allocator.dumpState(checkpoint);
+
+        int queueCounter = 0;
+        for (Map.Entry<QueueName, LinkedList<SegmentRef>> entry : queueSegments.entrySet()) {
+            // queues.0.name = bla bla
+            final QueueName queueName = entry.getKey();
+            checkpoint.setProperty("queues." + queueCounter + ".name", queueName.name);
+
+            // queues.0.segments = head (id_page, offset), (id_page, offset), ... tail
+            final LinkedList<SegmentRef> segmentRefs = entry.getValue();
+            final String segmentsDef = segmentRefs.stream()
+                .map(SegmentRef::toString)
+                .collect(Collectors.joining(", "));
+            checkpoint.setProperty("queues." + queueCounter + ".segments", segmentsDef);
+
+            // queues.0.head_offset = bytes offset from the start of the page where last data was written
+            final Queue queue = queues.get(queueName);
+            checkpoint.setProperty("queues." + queueCounter + ".head_offset", String.valueOf(queue.currentHead().segmentOffset(segmentSize)));
+            checkpoint.setProperty("queues." + queueCounter + ".tail_offset", String.valueOf(queue.currentTail().segmentOffset(segmentSize)));
+        }
+
+        final File propertiesFile = dataPath.resolve("checkpoint.properties").toFile();
+        final FileWriter fileWriter;
+        try {
+            fileWriter = new FileWriter(propertiesFile);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *fileWriter*. Execution paths that do not contain closure statements were detected. Ensure *fileWriter* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).